File: python/drift.py
"""
Drift SDK for Python Atomic functions.

This single-file SDK provides:
  - run(handler): Entry point that dispatches to deployed or local mode.
  - Backbone helpers: secret, cache, nosql, queue, blob, lock, vector.
  - log(msg): Writes to stderr (captured by the runner as function logs).
  - http_request(): Outbound HTTP from within a function.

All backbone helpers use only stdlib (urllib.request) -- zero external dependencies.
"""

import json
import os
import sys
import urllib.request
import urllib.parse
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler

# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def run(handler):
    """Entry point for Drift Atomic functions.

    In deployed mode (DRIFT_RUNTIME is set): reads JSON from stdin,
    calls handler, writes JSON to stdout.

    In local dev mode: starts an HTTP server on PORT (default 8080).
    """
    if os.environ.get("DRIFT_RUNTIME"):
        _run_deployed(handler)
    else:
        _run_local(handler)


def _run_deployed(handler):
    req = json.loads(sys.stdin.read())
    resp = handler(req)
    sys.stdout.write(json.dumps(resp))
    sys.stdout.flush()


def _run_local(handler):
    port = int(os.environ.get("PORT", "8080"))

    class Handler(BaseHTTPRequestHandler):
        def do_GET(self):
            self._handle()

        def do_POST(self):
            self._handle()

        def do_PUT(self):
            self._handle()

        def do_DELETE(self):
            self._handle()

        def _handle(self):
            content_length = int(self.headers.get("Content-Length", 0))
            body = None
            if content_length > 0:
                raw = self.rfile.read(content_length)
                try:
                    body = json.loads(raw)
                except json.JSONDecodeError:
                    body = raw.decode("utf-8", errors="replace")

            parsed = urllib.parse.urlparse(self.path)
            headers = {k: self.headers[k] for k in self.headers}

            req = {
                "method": self.command,
                "path": parsed.path,
                "headers": headers,
                "query": parsed.query,
                "body": body,
            }

            resp = handler(req)
            status = resp.get("status", 200)
            out = json.dumps(resp).encode("utf-8")

            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(out)

        def log_message(self, fmt, *args):
            sys.stderr.write(f"drift-sdk: {fmt % args}\n")

    server = HTTPServer(("", port), Handler)
    sys.stderr.write(f"drift-sdk: local server starting on :{port}\n")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass


# ---------------------------------------------------------------------------
# Backbone transport
# ---------------------------------------------------------------------------

_backbone_url = None


def _get_backbone_url():
    global _backbone_url
    if _backbone_url is None:
        _backbone_url = os.environ.get("BACKBONE_URL", "")
    return _backbone_url


def _call(method, path, body=None):
    """Call backbone via HTTP (deployed) or return None (local dev fallback)."""
    base = _get_backbone_url()
    if not base:
        return _call_local(method, path, body)

    url = f"{base}/{path}"
    data = None
    headers = {}
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["Content-Type"] = "application/json"

    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            raw = resp.read()
            if not raw:
                return None
            try:
                return json.loads(raw)
            except (json.JSONDecodeError, ValueError):
                return raw.decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        if e.code == 204:
            return None
        raise


# In-memory backbone for local dev (matches Go SDK behavior).
_local_store = {
    "nosql": {},
    "cache": {},
    "queues": {},
    "blobs": {},
    "locks": {},
    "next_id": 0,
}


def _call_local(method, path, body=None):
    """In-memory backbone for local development."""
    s = _local_store
    base_path = path.split("?")[0]
    query = {}
    if "?" in path:
        query = dict(urllib.parse.parse_qsl(path.split("?", 1)[1]))

    # NoSQL
    if base_path == "write" and method == "POST":
        col = (body or {}).get("collection", "default")
        if col not in s["nosql"]:
            s["nosql"][col] = {}
        s["next_id"] += 1
        key = str(s["next_id"])
        s["nosql"][col][key] = body
        return {"key": key}

    if base_path == "read" and method == "GET":
        col = query.get("collection", "default")
        key = query.get("key", "")
        return s["nosql"].get(col, {}).get(key)

    if base_path == "nosql/list" and method == "GET":
        col = query.get("collection", "default")
        docs = s["nosql"].get(col, {})
        field = query.get("field")
        value = query.get("value")
        results = []
        for doc in docs.values():
            if field and str(doc.get(field)) != value:
                continue
            results.append(doc)
        return results

    if base_path == "nosql/drop" and method == "POST":
        col = query.get("collection", "default")
        s["nosql"].pop(col, None)
        return None

    # Cache
    if base_path == "cache/set" and method == "POST":
        s["cache"][(body or {}).get("key", "")] = (body or {}).get("value")
        return None

    if base_path == "cache/get" and method == "GET":
        return s["cache"].get(query.get("key", ""))

    if base_path == "cache/del":
        s["cache"].pop(query.get("key", ""), None)
        return None

    # Queue
    if base_path == "queue/push" and method == "POST":
        name = (body or {}).get("queue", "")
        msg = (body or {}).get("body")
        s["queues"].setdefault(name, []).append(msg)
        return None

    if base_path == "queue/pop" and method == "POST":
        name = (body or {}).get("queue", "")
        q = s["queues"].get(name, [])
        if not q:
            return None
        return q.pop(0)

    # Blob
    if base_path == "blob/put" and method == "POST":
        s["blobs"][(body or {}).get("name", "")] = (body or {}).get("data")
        return None

    if base_path == "blob/get" and method == "GET":
        return s["blobs"].get(query.get("name", ""))

    # Secret
    if base_path == "secret/get" and method == "GET":
        return None

    # Lock
    if base_path == "lock/acquire" and method == "POST":
        name = (body or {}).get("name", "")
        if name in s["locks"]:
            return None
        s["next_id"] += 1
        token = f"local-lock-{s['next_id']}"
        s["locks"][name] = token
        return {"token": token}

    if base_path == "lock/release" and method == "POST":
        s["locks"].pop((body or {}).get("name", ""), None)
        return None

    # Vector
    if base_path == "vector/insert" and method == "POST":
        return None

    if base_path == "vector/search" and method == "POST":
        return []

    return None


# ---------------------------------------------------------------------------
# Secret
# ---------------------------------------------------------------------------

class _SecretNS:
    def get(self, name):
        resp = _call("GET", f"secret/get?name={urllib.parse.quote(name)}")
        return resp if isinstance(resp, str) else (json.dumps(resp) if resp else "")

    def set(self, name, value):
        _call("POST", "secret/set", {"name": name, "value": value})

    def delete(self, name):
        _call("DELETE", f"secret/delete?name={urllib.parse.quote(name)}")

secret = _SecretNS()


# ---------------------------------------------------------------------------
# Cache
# ---------------------------------------------------------------------------

class _CacheNS:
    def get(self, key):
        return _call("GET", f"cache/get?key={urllib.parse.quote(key)}")

    def set(self, key, value, ttl=0):
        payload = {"key": key, "value": value}
        if ttl > 0:
            payload["ttl"] = ttl
        _call("POST", "cache/set", payload)

    def delete(self, key):
        _call("DELETE", f"cache/del?key={urllib.parse.quote(key)}")

cache = _CacheNS()


# ---------------------------------------------------------------------------
# NoSQL
# ---------------------------------------------------------------------------

class _NoSQLNS:
    def collection(self, name):
        return _CollectionHandle(name)

class _CollectionHandle:
    def __init__(self, name):
        self.name = name

    def insert(self, doc):
        payload = {"collection": self.name}
        if isinstance(doc, dict):
            payload.update(doc)
        else:
            payload["data"] = doc
        resp = _call("POST", "write", payload)
        if isinstance(resp, dict):
            return resp.get("key", "")
        return ""

    def read(self, key):
        return _call("GET", f"read?collection={urllib.parse.quote(self.name)}&key={urllib.parse.quote(key)}")

    def list(self, filter=None):
        path = f"nosql/list?collection={urllib.parse.quote(self.name)}"
        if filter:
            for k, v in filter.items():
                path += f"&field={urllib.parse.quote(k)}&value={urllib.parse.quote(v)}"
        resp = _call("GET", path)
        return resp if isinstance(resp, list) else []

    def drop(self):
        _call("POST", f"nosql/drop?collection={urllib.parse.quote(self.name)}")

nosql = _NoSQLNS()


# ---------------------------------------------------------------------------
# Queue
# ---------------------------------------------------------------------------

class _QueueHandle:
    def __init__(self, name):
        self.name = name

    def push(self, body):
        _call("POST", "queue/push", {"queue": self.name, "body": body})

    def pop(self):
        return _call("POST", "queue/pop", {"queue": self.name})

def queue(name):
    return _QueueHandle(name)


# ---------------------------------------------------------------------------
# Blob
# ---------------------------------------------------------------------------

class _BlobNS:
    def put(self, name, data):
        _call("POST", "blob/put", {"name": name, "data": data})

    def get(self, name):
        return _call("GET", f"blob/get?name={urllib.parse.quote(name)}")

blob = _BlobNS()


# ---------------------------------------------------------------------------
# Lock
# ---------------------------------------------------------------------------

class _LockNS:
    def acquire(self, name, ttl=30):
        resp = _call("POST", "lock/acquire", {"name": name, "ttl": ttl})
        return (resp or {}).get("token", "")

    def release(self, name, token):
        _call("POST", "lock/release", {"name": name, "token": token})

lock = _LockNS()


# ---------------------------------------------------------------------------
# Vector
# ---------------------------------------------------------------------------

class _VectorNS:
    def collection(self, name):
        return _VectorCollectionHandle(name)

class _VectorCollectionHandle:
    def __init__(self, name):
        self.name = name

    def insert(self, id, vector, metadata=None):
        _call("POST", "vector/insert", {
            "collection": self.name,
            "id": id,
            "vector": vector,
            "metadata": metadata,
        })

    def search(self, vector, k=10):
        resp = _call("POST", "vector/search", {
            "collection": self.name,
            "vector": vector,
            "k": k,
        })
        return resp if isinstance(resp, list) else []

vector = _VectorNS()


# ---------------------------------------------------------------------------
# Logging
# ---------------------------------------------------------------------------

def log(msg):
    """Write a log message to stderr (captured by the runner)."""
    sys.stderr.write(str(msg) + "\n")
    sys.stderr.flush()


# ---------------------------------------------------------------------------
# HTTP client
# ---------------------------------------------------------------------------

def http_request(method, url, headers=None, body=None):
    """Make an outbound HTTP request. Returns (status, body_bytes)."""
    data = body if isinstance(body, bytes) else (body.encode("utf-8") if body else None)
    req = urllib.request.Request(url, data=data, headers=headers or {}, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            return resp.status, resp.read()
    except urllib.error.HTTPError as e:
        return e.code, e.read()

File: ruby/drift.rb
# Drift SDK for Ruby Atomic functions.
#
# This single-file SDK provides:
#   - Drift.run(handler): Entry point that dispatches to deployed or local mode.
#   - Backbone helpers: Secret, Cache, Nosql, Queue, Blob, Lock, Vector.
#   - Drift.log(msg): Writes to stderr (captured by the runner as function logs).
#   - Drift.http_request(): Outbound HTTP from within a function.
#
# All backbone helpers use only stdlib (net/http) -- zero external dependencies.

require 'json'
require 'net/http'
require 'uri'
require 'socket'

module Drift

  # ---------------------------------------------------------------------------
  # Entry point
  # ---------------------------------------------------------------------------

  def self.run(handler)
    if ENV['DRIFT_RUNTIME']
      _run_deployed(handler)
    else
      _run_local(handler)
    end
  end

  def self._run_deployed(handler)
    req = JSON.parse($stdin.read)
    resp = handler.call(req)
    $stdout.write(JSON.generate(resp))
    $stdout.flush
  end

  def self._run_local(handler)
    port = (ENV['PORT'] || '8080').to_i
    server = TCPServer.new('', port)
    $stderr.puts "drift-sdk: local server starting on :#{port}"

    loop do
      client = server.accept
      begin
        request_line = client.gets
        next unless request_line
        method, path_str, _ = request_line.strip.split(' ')

        headers = {}
        while (line = client.gets) && line.strip != ''
          key, value = line.strip.split(': ', 2)
          headers[key.downcase] = value if key
        end

        body = nil
        if headers['content-length']
          raw = client.read(headers['content-length'].to_i)
          begin
            body = JSON.parse(raw)
          rescue JSON::ParserError
            body = raw
          end
        end

        parsed = URI.parse(path_str || '/')
        req = {
          'method' => method,
          'path' => parsed.path,
          'headers' => headers,
          'query' => parsed.query || '',
          'body' => body,
        }

        resp = handler.call(req)
        status = resp['status'] || 200
        out = JSON.generate(resp)

        client.print "HTTP/1.1 #{status} OK\r\n"
        client.print "Content-Type: application/json\r\n"
        client.print "Content-Length: #{out.bytesize}\r\n"
        client.print "\r\n"
        client.print out
      rescue => e
        $stderr.puts "drift-sdk: #{e.message}"
      ensure
        client.close rescue nil
      end
    end
  end

  # ---------------------------------------------------------------------------
  # Backbone transport
  # ---------------------------------------------------------------------------

  @backbone_url = nil

  def self._get_backbone_url
    @backbone_url ||= ENV['BACKBONE_URL'] || ''
  end

  def self._call(method, path, body = nil)
    base = _get_backbone_url
    return _call_local(method, path, body) if base.empty?

    uri = URI("#{base}/#{path}")
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'

    request = case method
    when 'GET'    then Net::HTTP::Get.new(uri)
    when 'POST'   then Net::HTTP::Post.new(uri)
    when 'PUT'    then Net::HTTP::Put.new(uri)
    when 'DELETE' then Net::HTTP::Delete.new(uri)
    else Net::HTTP::Get.new(uri)
    end

    if body
      request['Content-Type'] = 'application/json'
      request.body = JSON.generate(body)
    end

    resp = http.request(request)
    return nil if resp.code == '204'
    return nil if resp.body.nil? || resp.body.empty?

    begin
      JSON.parse(resp.body)
    rescue JSON::ParserError
      resp.body
    end
  end

  # In-memory backbone for local dev.
  @local_store = {
    'nosql' => {}, 'cache' => {}, 'queues' => {},
    'blobs' => {}, 'locks' => {}, 'next_id' => 0,
  }

  def self._call_local(method, path, body = nil)
    s = @local_store
    base_path, qs = path.split('?', 2)
    query = {}
    if qs
      qs.split('&').each do |pair|
        k, v = pair.split('=', 2)
        query[URI.decode_www_form_component(k)] = URI.decode_www_form_component(v || '')
      end
    end

    # NoSQL
    if base_path == 'write' && method == 'POST'
      col = (body || {})['collection'] || 'default'
      s['nosql'][col] ||= {}
      s['next_id'] += 1
      key = s['next_id'].to_s
      s['nosql'][col][key] = body
      return { 'key' => key }
    end
    if base_path == 'read' && method == 'GET'
      col = query['collection'] || 'default'
      return (s['nosql'][col] || {})[query['key'] || '']
    end
    if base_path == 'nosql/list' && method == 'GET'
      col = query['collection'] || 'default'
      docs = s['nosql'][col] || {}
      field = query['field']
      value = query['value']
      results = []
      docs.each_value do |doc|
        next if field && doc[field].to_s != value
        results << doc
      end
      return results
    end
    if base_path == 'nosql/drop' && method == 'POST'
      s['nosql'].delete(query['collection'] || 'default')
      return nil
    end

    # Cache
    if base_path == 'cache/set' && method == 'POST'
      s['cache'][(body || {})['key'] || ''] = (body || {})['value']
      return nil
    end
    if base_path == 'cache/get' && method == 'GET'
      return s['cache'][query['key'] || '']
    end
    if base_path == 'cache/del'
      s['cache'].delete(query['key'] || '')
      return nil
    end

    # Queue
    if base_path == 'queue/push' && method == 'POST'
      name = (body || {})['queue'] || ''
      s['queues'][name] ||= []
      s['queues'][name] << (body || {})['body']
      return nil
    end
    if base_path == 'queue/pop' && method == 'POST'
      name = (body || {})['queue'] || ''
      q = s['queues'][name] || []
      return nil if q.empty?
      return q.shift
    end

    # Blob
    if base_path == 'blob/put' && method == 'POST'
      s['blobs'][(body || {})['name'] || ''] = (body || {})['data']
      return nil
    end
    if base_path == 'blob/get' && method == 'GET'
      return s['blobs'][query['name'] || '']
    end

    # Secret
    return nil if base_path == 'secret/get' && method == 'GET'

    # Lock
    if base_path == 'lock/acquire' && method == 'POST'
      name = (body || {})['name'] || ''
      return nil if s['locks'].key?(name)
      s['next_id'] += 1
      token = "local-lock-#{s['next_id']}"
      s['locks'][name] = token
      return { 'token' => token }
    end
    if base_path == 'lock/release' && method == 'POST'
      s['locks'].delete((body || {})['name'] || '')
      return nil
    end

    # Vector
    return nil if base_path == 'vector/insert' && method == 'POST'
    return [] if base_path == 'vector/search' && method == 'POST'

    nil
  end

  # ---------------------------------------------------------------------------
  # Secret
  # ---------------------------------------------------------------------------

  module Secret
    def self.get(name)
      resp = Drift._call('GET', "secret/get?name=#{URI.encode_www_form_component(name)}")
      resp.is_a?(String) ? resp : (resp ? JSON.generate(resp) : '')
    end

    def self.set(name, value)
      Drift._call('POST', 'secret/set', { 'name' => name, 'value' => value })
    end

    def self.delete(name)
      Drift._call('DELETE', "secret/delete?name=#{URI.encode_www_form_component(name)}")
    end
  end

  # ---------------------------------------------------------------------------
  # Cache
  # ---------------------------------------------------------------------------

  module Cache
    def self.get(key)
      Drift._call('GET', "cache/get?key=#{URI.encode_www_form_component(key)}")
    end

    def self.set(key, value, ttl: 0)
      payload = { 'key' => key, 'value' => value }
      payload['ttl'] = ttl if ttl > 0
      Drift._call('POST', 'cache/set', payload)
    end

    def self.delete(key)
      Drift._call('DELETE', "cache/del?key=#{URI.encode_www_form_component(key)}")
    end
  end

  # ---------------------------------------------------------------------------
  # NoSQL
  # ---------------------------------------------------------------------------

  module Nosql
    def self.collection(name)
      Collection.new(name)
    end

    class Collection
      def initialize(name)
        @name = name
      end

      def insert(doc)
        payload = { 'collection' => @name }
        if doc.is_a?(Hash)
          payload.merge!(doc)
        else
          payload['data'] = doc
        end
        resp = Drift._call('POST', 'write', payload)
        resp.is_a?(Hash) ? (resp['key'] || '') : ''
      end

      def read(key)
        Drift._call('GET', "read?collection=#{URI.encode_www_form_component(@name)}&key=#{URI.encode_www_form_component(key)}")
      end

      def list(filter = nil)
        path = "nosql/list?collection=#{URI.encode_www_form_component(@name)}"
        if filter
          filter.each do |k, v|
            path += "&field=#{URI.encode_www_form_component(k)}&value=#{URI.encode_www_form_component(v)}"
          end
        end
        resp = Drift._call('GET', path)
        resp.is_a?(Array) ? resp : []
      end

      def drop
        Drift._call('POST', "nosql/drop?collection=#{URI.encode_www_form_component(@name)}")
      end
    end
  end

  # ---------------------------------------------------------------------------
  # Queue
  # ---------------------------------------------------------------------------

  def self.queue(name)
    QueueHandle.new(name)
  end

  class QueueHandle
    def initialize(name)
      @name = name
    end

    def push(body)
      Drift._call('POST', 'queue/push', { 'queue' => @name, 'body' => body })
    end

    def pop
      Drift._call('POST', 'queue/pop', { 'queue' => @name })
    end
  end

  # ---------------------------------------------------------------------------
  # Blob
  # ---------------------------------------------------------------------------

  module Blob
    def self.put(name, data)
      Drift._call('POST', 'blob/put', { 'name' => name, 'data' => data })
    end

    def self.get(name)
      Drift._call('GET', "blob/get?name=#{URI.encode_www_form_component(name)}")
    end
  end

  # ---------------------------------------------------------------------------
  # Lock
  # ---------------------------------------------------------------------------

  module Lock
    def self.acquire(name, ttl: 30)
      resp = Drift._call('POST', 'lock/acquire', { 'name' => name, 'ttl' => ttl })
      (resp || {})['token'] || ''
    end

    def self.release(name, token)
      Drift._call('POST', 'lock/release', { 'name' => name, 'token' => token })
    end
  end

  # ---------------------------------------------------------------------------
  # Vector
  # ---------------------------------------------------------------------------

  module Vector
    def self.collection(name)
      VectorCollection.new(name)
    end

    class VectorCollection
      def initialize(name)
        @name = name
      end

      def insert(id, vector, metadata = nil)
        Drift._call('POST', 'vector/insert', {
          'collection' => @name, 'id' => id,
          'vector' => vector, 'metadata' => metadata,
        })
      end

      def search(vector, k: 10)
        resp = Drift._call('POST', 'vector/search', {
          'collection' => @name, 'vector' => vector, 'k' => k,
        })
        resp.is_a?(Array) ? resp : []
      end
    end
  end

  # ---------------------------------------------------------------------------
  # Logging
  # ---------------------------------------------------------------------------

  def self.log(msg)
    $stderr.puts msg.to_s
    $stderr.flush
  end

  # ---------------------------------------------------------------------------
  # HTTP client
  # ---------------------------------------------------------------------------

  def self.http_request(method, url, headers = {}, body = nil)
    uri = URI(url)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'

    request = case method.upcase
    when 'GET'    then Net::HTTP::Get.new(uri)
    when 'POST'   then Net::HTTP::Post.new(uri)
    when 'PUT'    then Net::HTTP::Put.new(uri)
    when 'DELETE' then Net::HTTP::Delete.new(uri)
    when 'PATCH'  then Net::HTTP::Patch.new(uri)
    else Net::HTTP::Get.new(uri)
    end

    (headers || {}).each { |k, v| request[k] = v }
    request.body = body if body

    resp = http.request(request)
    { 'status' => resp.code.to_i, 'body' => resp.body }
  end

end
